# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [2]:
import pandas as pd
import numpy as np
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv"
df = pd.read_csv(url)
df.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [3]:
columns = df.columns.str.strip().str.replace(' ', '_').str.lower()
df.columns = columns
#df.rename(columns={df.columns[1]: "state"}, inplace=True)

In [4]:
df["month"] = df["effective_to_date"].str.split("/").str[0]
df

unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       number_of_policies     policy_type        policy  renew_offer_type  \
0                       9  Corporate Auto  Corporate L3            Offer3   
1                       1   Personal Auto   Personal L3            Offer4   
2                       2   Personal Auto   Personal L3            Offer3   
3                       2  Corporate Auto  Corporate L3            Offer2   
4                       7   Personal Auto   Personal L2            Offer1   
...                   ...             ...           ...               ...   
10905                   7   Personal Auto   Personal L1            Offer3   
10906                   6   Personal Auto   Personal L3            Offer2   
10907                   2  Corporate Auto  Corporate L3            Offer1   
10908                   6   Personal Auto   Personal L1            Offer1   
10909                   3   Personal Auto   Personal L1            Offer4   

       sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0              Agent          292.800000  Four-Door Car       Medsize   
1        Call Center          744.924331  Four-Door Car       Medsize   
2        Call Center          480.000000            SUV       Medsize   
3             Branch          484.013411  Four-Door Car       Medsize   
4             Branch          707.925645  Four-Door Car       Medsize   
...              ...                 ...            ...           ...   
10905            Web         1214.400000     Luxury Car       Medsize   
10906         Branch          273.018929  Four-Door Car       Medsize   
10907            Web          381.306996     Luxury SUV       Medsize   
10908         Branch          618.288849            SUV       Medsize   
10909            Web         1021.719397            SUV       Medsize   

      vehicle_type month  
0              NaN     2  
1              NaN     1  
2                A     2  
3                A     1  
4              NaN     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907          NaN     2  
10908            A     2  
10909          NaN     1  

[10910 rows x 27 columns]

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [5]:
df_response_yes = df[(df['total_claim_amount'] > 1000) & (df['response'] == 'Yes')]
df_response_yes.head()

unnamed:_0 customer       state  customer_lifetime_value response  \
189         189  OK31456  California             11009.130490      Yes   
236         236  YJ16163      Oregon             11009.130490      Yes   
419         419  GW43195      Oregon             25807.063000      Yes   
442         442  IP94270     Arizona             13736.132500      Yes   
587         587  FJ28407  California              5619.689084      Yes   

     coverage             education effective_to_date employmentstatus gender  \
189   Premium              Bachelor           1/24/11         Employed      F   
236   Premium              Bachelor           1/24/11         Employed      F   
419  Extended               College           2/13/11         Employed      F   
442   Premium                Master           2/13/11         Disabled      F   
587   Premium  High School or Below           1/26/11       Unemployed      M   

     ...  number_of_policies     policy_type        policy  renew_offer_type  \
189  ...                   1  Corporate Auto  Corporate L3            Offer2   
236  ...                   1    Special Auto    Special L3            Offer2   
419  ...                   2   Personal Auto   Personal L2            Offer1   
442  ...                   8   Personal Auto   Personal L2            Offer1   
587  ...                   1   Personal Auto   Personal L1            Offer2   

     sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
189          Agent         1358.400000     Luxury Car       Medsize   
236          Agent         1358.400000     Luxury Car       Medsize   
419         Branch         1027.200000     Luxury Car         Small   
442            Web         1261.319869            SUV       Medsize   
587            Web         1027.000029            SUV       Medsize   

    vehicle_type month  
189          NaN     1  
236            A     1  
419            A     2  
442            A     2  
587            A     1  

[5 rows x 27 columns]

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [6]:
df_response_yes.pivot_table(values='total_claim_amount', index='policy_type', columns='gender', aggfunc='mean').round(2)

gender                F        M
policy_type                     
Corporate Auto  1138.40  1171.15
Personal Auto   1214.85  1137.86
Special Auto    1358.40  1017.50

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [9]:
# df["state"].value_counts()
df[df['state'].map(df['state'].value_counts()) > 500]


unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10904       10904  QI63521      Nevada             19228.463620       No   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   

       coverage             education effective_to_date employmentstatus  \
0         Basic               College           2/18/11         Employed   
1         Basic               College           1/18/11       Unemployed   
2         Basic              Bachelor           2/10/11         Employed   
3      Extended               College           1/11/11         Employed   
4       Premium              Bachelor           1/17/11    Medical Leave   
...         ...                   ...               ...              ...   
10904     Basic  High School or Below           2/24/11       Unemployed   
10905   Premium              Bachelor           1/19/11       Unemployed   
10906     Basic               College            1/6/11         Employed   
10907  Extended              Bachelor            2/6/11         Employed   
10908   Premium               College           2/13/11         Employed   

      gender  ...  number_of_policies     policy_type        policy  \
0          M  ...                   9  Corporate Auto  Corporate L3   
1          F  ...                   1   Personal Auto   Personal L3   
2          M  ...                   2   Personal Auto   Personal L3   
3          M  ...                   2  Corporate Auto  Corporate L3   
4          F  ...                   7   Personal Auto   Personal L2   
...      ...  ...                 ...             ...           ...   
10904      M  ...                   2   Personal Auto   Personal L2   
10905      F  ...                   7   Personal Auto   Personal L1   
10906      F  ...                   6   Personal Auto   Personal L3   
10907      F  ...                   2  Corporate Auto  Corporate L3   
10908      F  ...                   6   Personal Auto   Personal L1   

       renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
0                Offer3          Agent          292.800000  Four-Door Car   
1                Offer4    Call Center          744.924331  Four-Door Car   
2                Offer3    Call Center          480.000000            SUV   
3                Offer2         Branch          484.013411  Four-Door Car   
4                Offer1         Branch          707.925645  Four-Door Car   
...                 ...            ...                 ...            ...   
10904            Offer1         Branch          897.600000     Luxury SUV   
10905            Offer3            Web         1214.400000     Luxury Car   
10906            Offer2         Branch          273.018929  Four-Door Car   
10907            Offer1            Web          381.306996     Luxury SUV   
10908            Offer1         Branch          618.288849            SUV   

       vehicle_size vehicle_type month  
0           Medsize          NaN     2  
1           Medsize          NaN     1  
2           Medsize            A     2  
3           Medsize            A     1  
4           Medsize          NaN     1  
...             ...          ...   ...  
10904       Medsize            A     2  
10905       Medsize            A     1  
10906       Medsize            A  

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [75]:
df.groupby(['education', 'gender'])['customer_lifetime_value'].agg(['max', 'min', 'median']).round(2)

max      min   median
education            gender                            
Bachelor             F       73225.96  1904.00  5640.51
                     M       67907.27  1898.01  5548.03
College              F       61850.19  1898.68  5623.61
                     M       61134.68  1918.12  6005.85
Doctor               F       44856.11  2395.57  5332.46
                     M       32677.34  2267.60  5577.67
High School or Below F       55277.45  2144.92  6039.55
                     M       83325.38  1940.98  6286.73
Master               F       51016.07  2417.78  5729.86
                     M       50568.26  2272.31  5579.10

Men with "High School or Below" education have the highest median CLV (6,286.73), followed by women in the same category (6,039.55). This indicates that this demographic consistently generates significant value.

Women with "Doctor" education have the lowest median CLV (5,332.46), suggesting they bring lower value on average compared to other groups.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

In [92]:
df.pivot_table(values='number_of_policies', index='state', columns='month', aggfunc='sum', fill_value=0)

month          1     2
state                 
Arizona     3052  2864
California  5673  4929
Nevada      1493  1278
Oregon      4697  3969
Washington  1358  1225

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

In [93]:
total_policies_by_state = df.pivot_table(values='number_of_policies', index='state', aggfunc='sum').nlargest(3, 'number_of_policies').index
df.pivot_table(
    values='number_of_policies',
    index='state',
    columns='month',
    aggfunc='sum',
    fill_value=0).loc[total_policies_by_state]

month          1     2
state                 
California  5673  4929
Oregon      4697  3969
Arizona     3052  2864

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [102]:
melted_data = df[['sales_channel', 'response']]


melted_data = melted_data.melt(id_vars=['sales_channel'], value_vars=['response'], var_name='metric', value_name='value')


response_rate = melted_data[melted_data['value'] == 'Yes'].groupby('sales_channel').size() / melted_data.groupby('sales_channel').size() * 100


response_rate_df = response_rate.reset_index()
response_rate_df.columns = ['sales_channel', 'response_rate']


print(response_rate_df)

  sales_channel  response_rate
0         Agent      18.005339
1        Branch      10.787558
2   Call Center      10.322279
3           Web      10.885609
